src/core/tsi/ssl_transport_security.cc

```diff
@@ -568,21 +572,84 @@ static tsi_result ssl_ctx_use_private_key(SSL_CTX* context, const char* pem_key,
   EVP_PKEY* private_key = nullptr;
   BIO* pem;
   GPR_ASSERT(pem_key_size <= INT_MAX);
-  pem = BIO_new_mem_buf((void*)pem_key, static_cast<int>(pem_key_size));
-  if (pem == nullptr) return TSI_OUT_OF_RESOURCES;
-  do {
-    private_key = PEM_read_bio_PrivateKey(pem, nullptr, nullptr, (void*)"");
-    if (private_key == nullptr) {
-      result = TSI_INVALID_ARGUMENT;
-      break;
-    }
-    if (!SSL_CTX_use_PrivateKey(context, private_key)) {
-      result = TSI_INVALID_ARGUMENT;
-      break;
-    }
-  } while (0);
-  if (private_key != nullptr) EVP_PKEY_free(private_key);
-  BIO_free(pem);
+
+// BoringSSL does not have ENGINE_load_dynamic, ENGINE_by_id
+// support.
+#ifndef OPENSSL_IS_BORINGSSL
+  if (strncmp(pem_key, kSslEnginePrefix, strlen(kSslEnginePrefix)) == 0) {
+    ENGINE* engine;
```

Comment:
Can you split this function into engine and non-engine part since it's getting bigger?

Reply:
This is trying to reuse the existing method of using private key if key specified is not a engine key. So these are not exclusive and I would like to avoid duplicating code.

Reply:
I meant that this function can be split like
```
if (strncmp(pem_key, kSslEnginePrefix, strlen(kSslEnginePrefix)) == 0) {
  ssl_ctx_use_engine_private_key(...)
} else {
  ssl_ctx_use_regular_private_key(...)
}
```

Reply:
Changed.